corehq/ex-submodules/pillowtop/pillow/interface.py

```diff
@@ -100,28 +100,74 @@ def run(self):
         self.process_changes(since=self.get_last_checkpoint_sequence(), forever=True)
 
     def _update_checkpoint(self, change, context):
-        if change:
+        if change and context:
             updated = self.fire_change_processed_event(change, context)
         else:
             updated = self.checkpoint.touch(min_interval=CHECKPOINT_MIN_WAIT)
         if updated:
             self._record_checkpoint_in_datadog()
 
+    @property
+    def _should_process_in_chunks(self):
+        # chunked processing is only supported for pillows with single processor
+        return (len(self.processors) == 1
+            and getattr(self.processors[0], 'processor_chunk_size', 0) > 0)
+
     def process_changes(self, since, forever):
         """
-        Process changes from the changes stream.
+        Handle changes stream from the changes stream and pass it to processors
         """
         context = PillowRuntimeContext(changes_seen=0)
         try:
+            changes_chunk = []
             for change in self.get_change_feed().iter_changes(since=since or None, forever=forever):
                 if change:
                     context.changes_seen += 1
-                    self.process_with_error_handling(change, context)
+                    if self._should_process_in_chunks:
+                        changes_chunk.append(change)
+                        if len(changes_chunk) == self.processors[0].processor_chunk_size:
+                            self.process_chunk_with_error_handling(changes_chunk, context)
+                            changes_chunk = []
+                    else:
+                        self.process_with_error_handling(change, context)
                 else:
                     self._update_checkpoint(None, None)
+            self.process_chunk_with_error_handling(changes_chunk, context)
         except PillowtopCheckpointReset:
+            # finish processing any ramining chunk
+            self.process_chunk_with_error_handling(changes_chunk, context)
             self.process_changes(since=self.get_last_checkpoint_sequence(), forever=forever)
 
+    def process_chunk_with_error_handling(self, changes_chunk, context):
+        """
+        Passes given changes_chunk to the processor for chunked processing
+            If there is an exception in chunked processing, falls back
+            to passing changes one by one to the processor
+        """
+        if not changes_chunk:
+            return
+        try:
+            # chunked processing is supported if there is only one processor
+            result = self.processors[0].process_changes_chunk(self, changes_chunk)
+        except BulkPorcessingError as ex:
```

Comment:
If the bulk processor raises an error that is not related to any particular change, the pillow will just fall back to serial processing.